staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"fmt"
+	"net/http"
+
+	// TODO: decide whether to use the existing metrics, which
+	// categorize according to mutating vs readonly, or make new
+	// metrics because this filter does not pay attention to that
+	// distinction
+
+	// "k8s.io/apiserver/pkg/endpoints/metrics"
+
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	utilflowcontrol "k8s.io/apiserver/pkg/util/flowcontrol"
+	"k8s.io/klog"
+)
+
+// WithPriorityAndFairness limits the number of in-flight
+// requests in a fine-grained way.
+func WithPriorityAndFairness(
+	handler http.Handler,
+	longRunningRequestCheck apirequest.LongRunningRequestCheck,
+	fcIfc utilflowcontrol.Interface,
+) http.Handler {
+	if fcIfc == nil {
+		klog.Warningf("priority and fairness support not found, skipping")
+		return handler
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no RequestInfo found in context"))
+			return
+		}
+		user, ok := apirequest.UserFrom(ctx)
+		if !ok {
+			handleError(w, r, fmt.Errorf("no User found in context"))
+			return
+		}
+		requestDigest := utilflowcontrol.RequestDigest{requestInfo, user}
+
+		// Skip tracking long running requests.
+		if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
+			klog.V(6).Infof("Serving RequestInfo=%#+v, user.Info=%#+v as longrunning\n", requestInfo, user)
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		execute, afterExecute := fcIfc.Wait(ctx, requestDigest)
+		if !execute {
+			tooManyRequests(r, w)
+			return
+		}
+		defer afterExecute()
+		timedOut := ctx.Done()
+		finished := make(chan struct{})
+		go func() {
```

Comment:
Recent edits lost commentary here, but I have an open question here.  Should this goroutine do anything about recovering from panics, or anything like that?  If so, what?
@lavalamp

Reply:
Ah yeah, probably it does need the panic protection thingy.

Actually, what might be even better is if we catch any panic here, pass it through an error channel, and re-panic in the enclosing function. Otherwise it is very hard to debug them. I think the existing timeout handler contains this pattern; we should probably copy it.

Reply:
OK, you will see that in the next push.

Reply:
pushed yesterday